custom_cel/lists.go

```diff
@@ -0,0 +1,220 @@
+
+// CompileOptions implements the Library interface method defining the basic compile configuration
+func (u listsLib) CompileOptions() []cel.EnvOption {
+	dynListType := cel.ListType(cel.DynType)
+	sortByMacro := parser.NewReceiverMacro("sort_by", 2, makeSortBy)
+	sortByMacroWithOrder := parser.NewReceiverMacro("sort_by", 3, makeSortBy)
+	return []cel.EnvOption{
+		ext.Strings(),
+		library.Lists(),
+		cel.Macros(sortByMacro),
+		cel.Macros(sortByMacroWithOrder),
+		cel.Function(
+			"pair",
+			cel.Overload(
+				"make_pair",
+				[]*cel.Type{cel.DynType, cel.DynType},
+				cel.DynType,
+				cel.BinaryBinding(makePair),
+			),
+		),
+		cel.Function(
+			"sort",
+			cel.Overload(
+				"sort_by_order",
+				[]*cel.Type{dynListType, cel.DynType},
+				dynListType,
+				cel.BinaryBinding(sortByOrder),
+			),
+		),
+	}
+}
+
+// ProgramOptions implements the Library interface method defining the basic program options
+func (u listsLib) ProgramOptions() []cel.ProgramOption {
+	return []cel.ProgramOption{}
+}
+
+type pair struct {
+	order ref.Val
+	value ref.Val
+}
+
+var (
+	orderKey = types.DefaultTypeAdapter.NativeToValue("order")
+	valueKey = types.DefaultTypeAdapter.NativeToValue("value")
+)
+
+func makePair(order ref.Val, value ref.Val) ref.Val {
+	if _, ok := order.(traits.Comparer); !ok {
+		return types.ValOrErr(order, "unable to build ordered pair with value %v", order.Value())
+	}
+	return types.NewStringInterfaceMap(types.DefaultTypeAdapter, map[string]any{
+		"order": order.Value(),
+		"value": value.Value(),
+	})
+}
+
+func sortByOrder(itemsVal ref.Val, orderVal ref.Val) ref.Val {
+	items, ok := itemsVal.(traits.Lister)
+	if !ok {
+		return types.ValOrErr(itemsVal, "unable to convert to traits.Lister")
+	}
+
+	order, ok := orderVal.Value().(string)
+	if !ok {
+		return types.ValOrErr(orderVal, "unable to convert to ref.Val string")
+	}
+
+	pairs := make([]pair, 0, items.Size().Value().(int64))
+	index := 0
+	for it := items.Iterator(); it.HasNext().(types.Bool); {
+		curr, ok := it.Next().(traits.Mapper)
+		if !ok {
+			return types.NewErr("unable to convert elem %d to traits.Mapper", index)
+		}
+
+		pairs = append(pairs, pair{
+			order: curr.Get(orderKey),
+			value: curr.Get(valueKey),
+		})
+		index++
+	}
+
+	ascSort := func(i, j int) bool {
+		cmp := pairs[i].order.(traits.Comparer)
+		switch cmp.Compare(pairs[j].order) {
+		case types.IntNegOne:
+			return true
+		case types.IntOne:
+			return false
+		default: // IntZero means equal
+			return false
+		}
+	}
+
+	descSort := func(i, j int) bool {
+		cmp := pairs[i].order.(traits.Comparer)
+		switch cmp.Compare(pairs[j].order) {
+		case types.IntNegOne:
+			return false
+		case types.IntOne:
+			return true
+		default: // IntZero means equal
+			return false
+		}
+	}
+
+	switch strings.ToLower(order) {
+	case AscendingOrder:
+		sort.Slice(pairs, ascSort)
+	case DescendingOrder:
+		sort.Slice(pairs, descSort)
+	default:
+		return types.NewErr("unknown order: %s", order)
+	}
```

Comment:
Do we need to support "asc" and "desc"? We can drastically shave the code we have to maintain and scenarios we have to test (currently "asc", "desc" and arg omitted)

`sort_by(desc)` is also equivalent to `values.sort_by(i, -i)` or `values.sort_by(i, i).reverse()` (we'd need to add a reverse member func to dynList)

```suggestion
	cmp := func(i, j int) bool {
		return pairs[i].order.(traits.Comparer).Compare(pairs[j].order) == types.IntNegOne
	}
	sort.Slice(pairs, cmp)
```

Reply:
I think we should support a way to get both a list in ascending or descending order (consider, for example, the case where we want to retrieve the last releases or the first releases for example). About how we can do it, I think your idea about defining a `.reverse()` (in our implementation named `reverse_list` due to conflicts with `cel-go` source code) is a more "CEL-like" way to achieve it. So, implemented in cd4cc7f1e2adb9dfafafa38905ba9245c1a94186.